File: ios/AeSMS/Services/EventStream.swift
import Foundation

/// Online push: long-lived SSE to /v1/events. Triggers inbox refresh when a message lands.
actor EventStream {
    private var task: Task<Void, Never>?

    func start(base: String, token: String, onMessage: @escaping @Sendable () -> Void) {
        stop()
        task = Task {
            while !Task.isCancelled {
                do {
                    try await listen(base: base, token: token, onMessage: onMessage)
                } catch {
                    try? await Task.sleep(nanoseconds: 2_000_000_000)
                }
            }
        }
    }

    func stop() {
        task?.cancel()
        task = nil
    }

    private func listen(base: String, token: String, onMessage: @escaping @Sendable () -> Void) async throws {
        let trimmed = base.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard let url = URL(string: trimmed + "/v1/events") else { return }
        var req = URLRequest(url: url)
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        req.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        req.timeoutInterval = 3600

        let (bytes, resp) = try await URLSession.shared.bytes(for: req)
        guard let http = resp as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw URLError(.badServerResponse)
        }

        for try await line in bytes.lines {
            if Task.isCancelled { return }
            if line.hasPrefix("event: message") || line.contains("\"event\":\"message\"") {
                onMessage()
            }
        }
    }
}

File: android/app/src/main/java/io/aesms/app/data/EventStream.kt
package io.aesms.app.data

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okio.BufferedSource
import java.util.concurrent.TimeUnit

/** Long-lived SSE to /v1/events — refresh inbox when a message lands. */
class EventStream {
    private val client = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .connectTimeout(30, TimeUnit.SECONDS)
        .build()

    private var job: Job? = null

    fun start(scope: CoroutineScope, base: String, token: String, onMessage: () -> Unit) {
        stop()
        job = scope.launch(Dispatchers.IO) {
            while (isActive) {
                try {
                    listen(base, token, onMessage)
                } catch (_: Exception) {
                    delay(2_000)
                }
            }
        }
    }

    fun stop() {
        job?.cancel()
        job = null
    }

    private fun listen(base: String, token: String, onMessage: () -> Unit) {
        ApiClient().assertTransitSafe(base)
        val url = base.trimEnd('/') + "/v1/events"
        val req = Request.Builder()
            .url(url)
            .header("Authorization", "Bearer $token")
            .header("Accept", "text/event-stream")
            .build()
        client.newCall(req).execute().use { resp ->
            if (!resp.isSuccessful) error("bad SSE response")
            val source = resp.body?.source() ?: return
            readLines(source) { line ->
                if (line.startsWith("event: message") || line.contains("\"event\":\"message\"")) {
                    onMessage()
                }
            }
        }
    }

    private fun readLines(source: BufferedSource, onLine: (String) -> Unit) {
        while (!source.exhausted()) {
            val line = source.readUtf8Line() ?: break
            onLine(line)
        }
    }
}
